src/ast/ast.h

```diff
@@ -216,6 +216,19 @@ class Statement : public VariantNode<ExprStatement,
 };
 using StatementList = std::vector<Statement>;
 
+class Macro;
+class MapDeclStatement;
+class Probe;
+class Subprog;
+
+class RootStatement
+    : public VariantNode<Probe, Subprog, Macro, MapDeclStatement> {
+public:
+  using VariantNode::VariantNode;
+  RootStatement() : RootStatement(static_cast<Probe *>(nullptr)) {};
+};
+using RootStatements = std::vector<RootStatement>;
```

Comment:
I don't specifically have a program with this construction, but it's a bit weird because `RootStatement` is purely a byproduct of the parser AFAICT.

One thought that I had in the past is factoring the driver/context to hold a program as a non-pointer reference, e.g.

```
class ASTContext {
   ...
   Program root;
   ...
}
```

And then just have the parser add to the various blocks as it encounters the top-level statements, rather than needing the intermediate node for Bison. This also solves the issue of whether root is `nullptr`.

I'm good to approve as is, I'm just raising an alternative that you may want to consider.

Reply:
> factoring the driver/context to hold a program as a non-pointer reference

Ah neat. Let me look into that in a follow-up PR.